### PR TITLE
CNN Trainer Updates

### PR DESCRIPTION
**Pull request by Byron-Selvage**
# Summary
This PR implements three improvements to the CNN trainer:
1. Exposes validation loss
2. Adds early stopping criteria
3. Adds warmup to the LR schedule

These changes make the trainer more understandable, reduce overfitting risk, and make optimization more stable across hyperparameter settings.

# 1. Expose Validation Loss
The CNN trainer now computes and logs the validation loss in addition to training loss each epoch.

## What changed
- The trainer base class in `trainer.py` now passes held-out test data (from test-train split) into `inner_train`.
- The CNN trainer logs both training loss and validation loss each epoch.
- Other trainer implementations have been edited to cleanly accept the test data for interface compatibility but currently do not use it.

## Why
- Allows us to check if training and validation loss are trending together or diverging. 
- Enables earlier detection of overfitting.
- Provides the signal used for early stopping criteria.

# 2. Early Stopping Criteria
The CNN trainer now has optional early stopping based on the validation loss curve. Early stopping is off by default.

## New parameters
- `early_stopping_patience` (default `None`)
- `early_stopping_min_delta` (default `1e-6`)

## Behavior
When early stopping is enabled (stopping patience is not None), the Trainer:
- Tracks best validation loss.
- Saves the best model weights when the validation loss improves by more than `early_stopping_min_delta`.
- Stops when the validation loss does not improve for `early_stopping_patience` epochs.
- Restores the best model weights before returning.

## Why
Using early stopping:
- Prevents computational waste
- Prevents overfitting

# 3. LR Schedule Warmup
The CNN trainer now supports an optional LR warmup phase before cosine annealing. Warmup is off by default.

## New parameters
- `warmup_epochs` (default `0`)

## Behavior
- If `warmup_epochs > 0`, training uses linear warmup first, then cosine annealing.
- If `warmup_epochs = 0`, training uses cosine annealing only.

## Why
LR Warmup improves performance and makes hyperparameter tuning more robust. For reference, see the following: 
Kalra, Dayal Singh, and Maissam Barkeshli. _Why Warmup the Learning Rate? Underlying Mechanisms and Improvements._ November 1, 2024. https://doi.org/10.48550/arXiv.2406.09405.

### CODE REVIEW

**Review by nreinicke — CHANGES_REQUESTED** [review]
I like where this is going! I also like the idea of including a validation set but I think we should be a bit more explicit about it. Right now we're using the test split for both validation and error reporting which introduces a bias. What do you think about making the validation set a fixture of the trainer? For example, instead of doing a train/test split, we do a train/test/validation split and pass the train/validation datasets into the `inner_train` method. That way, the test data stays independent but we still provide a mechanism to make decisions based on the validation set. We could do something like 80/10/10 for a default ratio but let it be configurable.

**Comment by Byron-Selvage** [comment]
Great point Nick! To address this bias, in 75cf201 I've replaced the previous test/train split helper in utils.py with a train/validation/test split and updated the Trainer so that test data is fully held out during training. Specifically, `inner_train` now receives the train data (and validation data if the trainer supports it) while errors are computed only on the held-out test data. 

For default split amounts, I added trainer-level split defaults in the base Trainer class so each trainer can define its own behavior while still allowing split values to be configurable via the ModelConfig. 
- CNN defaults to an 80/10/10 split (train/validation/test).
- Random Forest and NGBoost default to an 80/20 (train/test) split with no validation set. 
 
This preserves the existing non-CNN behavior, removes the CNN bias issue, and still supports full configurability via the ModelConfig.

File: routee/powertrain/trainers/cnn.py [hunk 1]
```diff
@@ -185,12 +220,25 @@ def _train_on(dev: "torch.device"):
...
+                T_max=cosine_steps,
                 eta_min=self.min_learning_rate,
             )
+            if warmup_steps > 0:
+                warmup_scheduler = torch.optim.lr_scheduler.LinearLR(
+                    optimizer,
+                    start_factor=1e-8,
+                    total_iters=warmup_steps,
+                )
+                scheduler = torch.optim.lr_scheduler.ChainedScheduler(
```
Comment (nreinicke) [inline routee/powertrain/trainers/cnn.py]: Can you verify the intention for this scheduler? I'm wondering if the SequentialLR might be more suitable if the intent is to run linear warm up and _then_ run the cosine annealing? I think the ChainedScheduler might run both at the same time?

**Review by nreinicke — COMMENTED** [review]
Thanks for making these changes! I have one more question but overall this is looking good.

File: routee/powertrain/trainers/cnn.py (hunk 1 above)
Comment (Byron-Selvage) [reply routee/powertrain/trainers/cnn.py]: Well that's embarrassing. I fundamentally misunderstood how the PyTorch chained LR scheduler works. As you mentioned, the intention is to run the linear warmup and _then_ cosine annealing. With the ChainedScheduler however, both learning schedules are called every step. Here's a comparison between ChainedScheduler as implemented and SequentialLR.

Schedulers

The differences between the two schedules are small. However, ChainedScheduler is obviously creating unintended behavior (e.g. the LR begins to increase again at the end of the  schedule). 

To fix this, I have switched to using SequentialLR, which produces the intended scheduler behavior. Thank you very much for pointing this out!

**Review by nreinicke — APPROVED** [review]
Thanks for confirming and updating! This looks great, thanks for the additions!
